subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -889,20 +859,277 @@ static int nrf_cloud_encode_service_info_ui(const struct nrf_cloud_svc_info_ui *
+
+	ret = add_modem_info_data(&network->current_band, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&network->sup_band, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&network->area_code, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&network->current_operator, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&network->ip_address, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&network->ue_mode, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = modem_info_name_get(network->cellid_hex.type, data_name);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (cJSON_AddNumberToObjectCS(json_obj, data_name, network->cellid_dec) == NULL) {
+		return -EINVAL;
+	}
+
+	if (network->lte_mode.value == 1) {
+		strcat(network_mode, "LTE-M");
+	} else if (network->nbiot_mode.value == 1) {
+		strcat(network_mode, "NB-IoT");
+	}
+	if (network->gps_mode.value == 1) {
+		strcat(network_mode, " GPS");
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "networkMode", network_mode) == NULL) {
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int encode_modem_info_sim(struct sim_param *sim, cJSON *json_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(sim != NULL);
+	__ASSERT_NO_MSG(json_obj != NULL);
+
+	ret = add_modem_info_data(&sim->uicc, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_SIM_ICCID)) {
+		ret = add_modem_info_data(&sim->iccid, json_obj);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_SIM_IMSI)) {
+		ret = add_modem_info_data(&sim->imsi, json_obj);
+	}
+
+	return ret;
+}
+
+static int encode_modem_info_device(struct device_param *device, cJSON *json_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(device != NULL);
+	__ASSERT_NO_MSG(json_obj != NULL);
+
+	ret = add_modem_info_data(&device->modem_fw, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&device->battery, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	ret = add_modem_info_data(&device->imei, json_obj);
+	if (ret) {
+		return ret;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "board", device->board) == NULL) {
+		return -EINVAL;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "appVersion", device->app_version) == NULL) {
+		return -EINVAL;
+	}
+
+	if (cJSON_AddStringToObject(json_obj, "appName", device->app_name) == NULL) {
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int encode_modem_info_json_object(struct modem_param_info *modem,
+				  cJSON *root_obj)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(root_obj != NULL);
+	__ASSERT_NO_MSG(modem != NULL);
+
+	int obj_count = cJSON_GetArraySize(root_obj);
+
+	cJSON *network_obj = NULL;
+	cJSON *sim_obj = NULL;
+	cJSON *device_obj = NULL;
+
+	if (IS_ENABLED(CONFIG_MODEM_INFO_ADD_NETWORK)) {
+		network_obj = cJSON_CreateObject();
+		if (network_obj == NULL) {
+			return -ENOMEM;
+		}
+		ret = encode_modem_info_network(&modem->network, network_obj);
```

Comment:
If this fails, the cjson object needs to be deleted.
Also in the similar cases below.

Reply:
Good point! Address these cases now.